genesyscloud/resource_genesyscloud_integration.go

```diff
@@ -0,0 +1,366 @@
+			},
+			"advanced": {
+				Description:      "Integration advanced config (JSON string).",
+				Type:             schema.TypeString,
+				Optional:         true,
+				Computed:         true,
+				DiffSuppressFunc: suppressEquivalentJsonDiffs,
+			},
+			"credentials": {
+				Description: "Credentials required for the integration. The required keys are indicated in the credentials property of the Integration Type.",
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+)
+
+func getAllIntegrations(ctx context.Context, clientConfig *platformclientv2.Configuration) (ResourceIDMetaMap, diag.Diagnostics) {
+	resources := make(ResourceIDMetaMap)
+	integrationAPI := platformclientv2.NewIntegrationsApiWithConfig(clientConfig)
+
+	for pageNum := 1; ; pageNum++ {
+		integrations, _, err := integrationAPI.GetIntegrations(100, pageNum, "", nil, "", "")
+		if err != nil {
+			return nil, diag.Errorf("Failed to get page of integrations: %v", err)
+		}
+
+		if integrations.Entities == nil || len(*integrations.Entities) == 0 {
+			break
+		}
+
+		for _, integration := range *integrations.Entities {
+			resources[*integration.Id] = &ResourceMeta{Name: *integration.Name}
+		}
+	}
+
+	return resources, nil
+}
+
+func integrationExporter() *ResourceExporter {
+	return &ResourceExporter{
+		GetResourcesFunc: getAllWithPooledClient(getAllIntegrations),
+		RefAttrs: map[string]*RefAttrSettings{
+			"config.credentials.*": {}, // Ref type not yet defined
+		},
+	}
+}
+
+func resourceIntegration() *schema.Resource {
+	return &schema.Resource{
+		Description: "Genesys Cloud Integration",
+
+		CreateContext: createWithPooledClient(createIntegration),
+		ReadContext:   readWithPooledClient(readIntegration),
+		UpdateContext: updateWithPooledClient(updateIntegration),
+		DeleteContext: deleteWithPooledClient(deleteIntegration),
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"intended_state": {
+				Description:  "Integration state (ENABLED | DISABLED | DELETED).",
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "DISABLED",
+				ValidateFunc: validation.StringInSlice([]string{"ENABLED", "DISABLED", "DELETED"}, false),
+			},
+			"integration_type": {
+				Description: "Integration type.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"config": {
+				Description: "Integration config. Each integration type has different schema, use [GET /api/v2/integrations/types/{typeId}/configschemas/{configType}](https://developer.mypurecloud.com/api/rest/v2/integrations/#get-api-v2-integrations-types--typeId--configschemas--configType-) to check schema, then use the correct attribute names for properties.",
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				Optional:    true,
```

Comment:
Since this contains required attributes like name, this block should also be required now.

Reply:
I thought it's more like config attribute is optional, but when there is a config attr, then within it the name is required. I checked the user resource and found some [examples](https://github.com/MyPureCloud/terraform-provider-genesyscloud/blob/7cab741ec8778433fff441c7fa4bfe5341411f03/genesyscloud/resource_genesyscloud_user.go#L26) of that scenario.  Or I can just change the name to be optional if that doesn't work. Since users can still make request without specifying the name. The API will automatically assign a name anyway. The name attribute will be set when users want to change the name of the integration.

Reply:
Ah, the API docs say the name is required, but I guess that must be wrong: https://developer.genesys.cloud/api/rest/v2/integrations/#post-api-v2-integrations

We should add a test case without the config block set to confirm that works. Also if the API can set the name, the `name` attribute should be `Optional` and `Computed` so it can be excluded from the config if someone just wants to set `properties`

Reply:
Made changes to the `name` attribute and also refactored the `updateIntegrationConfig` function a little bit. Added two more test cases, one with no config block and one with config block but all nullValues.